### PR TITLE
FAN Zhen

### PR DESCRIPTION
**Pull request by DDDAZ**

### CODE REVIEW

File: EFCoreRelationshipsPractice/Services/CompanyService.cs [hunk 1]
```diff
@@ -18,22 +20,53 @@ public CompanyService(CompanyDbContext companyDbContext)
...
+            var companyEntity = new CompanyEntity(companyDto);
+            await companyDbContext.Companies.AddRangeAsync(companyEntity);
+            await companyDbContext.SaveChangesAsync();
+            return companyEntity.Id;
         }
 
         public async Task DeleteCompany(int id)
         {
-            throw new NotImplementedException();
+            var foundCompany = await companyDbContext.Companies
```
Comment (fuywingfuy) [inline EFCoreRelationshipsPractice/Services/CompanyService.cs:49]: good practice

File: EFCoreRelationshipsPractice/Services/CompanyService.cs [hunk 2]
```diff
@@ -18,22 +20,53 @@ public CompanyService(CompanyDbContext companyDbContext)
...
+                var foundEmployee = await companyDbContext.Employees.FirstOrDefaultAsync(employee => employee.Id == employeeEntity.Id);
+                if (foundEmployee != null)
+                {
+                    companyDbContext.Employees.Remove(foundEmployee);
+                }
+            }
+
+            companyDbContext.Companies.Remove(foundCompany);
+
+            await companyDbContext.SaveChangesAsync();
```
Comment (zxubootcamp) [inline EFCoreRelationshipsPractice/Services/CompanyService.cs:69]: Cool! Manaully implemented full delete logic, searching for all related entities.

File: EFCoreRelationshipsPracticeTest/ControllerTest/CompanyControllerTest.cs [hunk 3]
```diff
@@ -22,33 +20,25 @@ public CompanyControllerTest(CustomWebApplicationFactory<Startup> factory) : bas
...
-                }
-            };
-
-            companyDto.Profile = new ProfileDto()
-            {
-                RegisteredCapital = 100010,
-                CertId = "100",
+                Name = "IBM",
+                Employees = new List<EmployeeDto>() { new EmployeeDto() { Name = "Tom", Age = 19 } },
+                Profile = new ProfileDto() { RegisteredCapital = 100010, CertId = "100", },
```
Comment (zxubootcamp) [inline EFCoreRelationshipsPracticeTest/ControllerTest/CompanyControllerTest.cs:29]: Good reformat.

File: EFCoreRelationshipsPractice/Services/CompanyService.cs [hunk 4]
```diff
@@ -18,22 +20,53 @@ public CompanyService(CompanyDbContext companyDbContext)
...
+            foreach (var employeeEntity in foundCompany.Employees)
+            {
+                var foundEmployee = await companyDbContext.Employees.FirstOrDefaultAsync(employee => employee.Id == employeeEntity.Id);
+                if (foundEmployee != null)
+                {
+                    companyDbContext.Employees.Remove(foundEmployee);
+                }
+            }
+
+            companyDbContext.Companies.Remove(foundCompany);
```
Comment (YangJK2020) [inline EFCoreRelationshipsPractice/Services/CompanyService.cs:67]: Now we know a better solution is using foreign key. Good practice though

File: EFCoreRelationshipsPracticeTest/ControllerTest/CompanyControllerTest.cs [hunk 5]
```diff
@@ -1,18 +1,16 @@
...
 using System.Text;
 using System.Threading.Tasks;
-using EFCoreRelationshipsPractice;
-using EFCoreRelationshipsPractice.Dtos;
-using Newtonsoft.Json;
 using Xunit;
 
 namespace EFCoreRelationshipsPracticeTest
 {
+    [Collection("TestClass1")]
```
Comment (zxubootcamp) [inline EFCoreRelationshipsPracticeTest/ControllerTest/CompanyControllerTest.cs:13]: Nice, used collection to group test.

**Pull request closed without merging**